src/extensions/sort.rs

```diff
@@ -0,0 +1,148 @@
+//! Adds support for the IMAP SORT extension specificed in [RFC
+//! 5464](https://tools.ietf.org/html/rfc5256#section-3).
+//!
+//! The SORT command is a variant of SEARCH with sorting semantics for
+//! the results. There are two arguments before the searching
+//! criteria argument: a parenthesized list of sort criteria, and the
+//! searching charset.
+
+use std::{borrow::Cow, fmt};
+
+pub(crate) struct SortCriteria<'c>(pub(crate) &'c [SortCriterion<'c>]);
+
+impl<'c> fmt::Display for SortCriteria<'c> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if self.0.is_empty() {
+            write!(f, "")
+        } else {
+            let criteria: Vec<String> = self.0.iter().map(|c| c.to_string()).collect();
+            write!(f, "({})", criteria.join(" "))
+        }
+    }
+}
+
+/// The defined sort criteria are as follows. Refer to the Formal
+/// Syntax section for the precise syntactic definitions of the
+/// arguments. If the associated [RFC-822](https://tools.ietf.org/html/rfc822)
+/// header for a particular criterion is absent, it is treated as the empty string.
+/// The empty string always collates before non-empty strings.
+pub enum SortCriterion<'c> {
+    /// Internal date and time of the message. This differs from the
+    /// ON criteria in SEARCH, which uses just the internal date.
+    Arrival,
+
+    /// [IMAP](https://tools.ietf.org/html/rfc5256#ref-IMAP) addr-mailbox
+    /// of the first "cc" address.
+    Cc,
+
+    /// Sent date and time, as described in
+    /// [section 2.2](https://tools.ietf.org/html/rfc5256#section-2.20).
+    Date,
+
+    /// [IMAP](https://tools.ietf.org/html/rfc5256#ref-IMAP) addr-mailbox
+    /// of the first "From" address.
+    From,
+
+    /// Followed by another sort criterion, has the effect of that
+    /// criterion but in reverse (descending) order.
+    Reverse(&'c SortCriterion<'c>),
+
+    /// Size of the message in octets.
+    Size,
+
+    /// Base subject text.
+    Subject,
+
+    /// [IMAP](https://tools.ietf.org/html/rfc5256#ref-IMAP) addr-mailbox
+    /// of the first "To" address.
```

Comment:
Can we make this documentation a little clearer for someone who doesn't know the exact spec? What is an `addr-mailbox`?

Reply:
:heavy_check_mark: Done